yajsapi/storage/ws-browser.ts

```diff
@@ -0,0 +1,234 @@
+import { StorageProvider, StorageProviderDataCallback } from "./provider";
+import { v4 } from "uuid";
+import { encode, toObject } from "flatbuffers/js/flexbuffers.js";
+import { getIdentity } from "../network/identity.js";
+import * as jsSha3 from "js-sha3";
+import { Logger, nullLogger } from "../utils/index.js";
+
+
+export interface WebSocketStorageProviderOptions {
+  yagnaOptions: {
+    apiKey: string;
+    basePath: string;
+  };
+  logger?: Logger
+}
+
+interface GsbRequest<T> {
+  id: string;
+  component: string;
+  payload: T;
+}
+
+interface GetMetadataRequest extends GsbRequest<void> {
+  component: "GetMetadata";
+}
+
+interface GetChunkRequest extends GsbRequest<{ offset: number, size: number }> {
+  component: "GetChunk";
+}
+
+type UploadChunkChunk = { offset: number, content: Uint8Array }
+type UploadChunkPayload = {
+  chunk: UploadChunkChunk;
+};
+
+interface UploadChunkRequest extends GsbRequest<UploadChunkPayload> {
+  component: "UploadChunk";
+}
+
+interface UploadFinishedRequest extends GsbRequest<{ hash: string }> {
+  component: "UploadFinished"
+}
+
+type GsbRequestPublishUnion = GetMetadataRequest | GetChunkRequest;
+type GsbRequestReceiveUnion = UploadFinishedRequest | UploadChunkRequest;
+
+type ServiceInfo = {
+  url: URL;
+  serviceId: string;
+}
+
+type GftpFileInfo = {
+  id: string;
+  url: string;
+}
+
+/**
+ * Storage provider that uses GFTP over WebSockets.
+ */
+export class WebSocketBrowserStorageProvider implements StorageProvider {
+  /**
+   * Map of open services (IDs) indexed by GFTP url.
+   */
+  private services = new Map<string, string>();
+  private logger: Logger;
+
+  constructor(private readonly options: WebSocketStorageProviderOptions) {
+    this.logger = options.logger ?? nullLogger();
```

Comment:
Is it necessary to use null Logger? we haven't done that before and the logger was always nullable..

Reply:
Having null objects for complex object with logic make the code cleaner, you don't need to check for it's existence every time you use it and the code can simply assume it's always there.

Null logger was introduced in logger improvements, and similarly null storage provider was introduced in storage provider improvements (though probably not really a null object itself)